workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ComposeLifecycleOwner.kt

```diff
@@ -0,0 +1,75 @@
+package com.squareup.workflow1.ui.compose
+
+import androidx.compose.runtime.RememberObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LifecycleRegistry
+
+/**
+ * A custom [LifecycleOwner] that synchronizes its lifecycle with a parent [Lifecycle] and
+ * integrates with Jetpack Compose's lifecycle through [RememberObserver].
+ *
+ * **Purpose:**
+ *
+ * - Ensures that any lifecycle-aware components within a composable function have a lifecycle that
+ *   accurately reflects both the parent lifecycle and the composable's own lifecycle.
+ * - Manages lifecycle transitions and observer registration/removal to prevent memory leaks and
+ *   ensure proper cleanup when the composable leaves the composition.
+ *
+ * **Key Features:**
+ *
+ * - **Lifecycle Synchronization:** Mirrors lifecycle events from the provided `parentLifecycle` to
+ *   its own [LifecycleRegistry], ensuring consistent state transitions.
+ * - **Compose Integration:** Implements [RememberObserver] to align with the composable's lifecycle
+ *   in the Compose memory model.
+ * - **Automatic Observer Management:** Adds and removes a [LifecycleEventObserver] to the parent
+ *   lifecycle, preventing leaks and ensuring proper disposal.
+ * - **State Transition Safety:** Carefully manages lifecycle state changes to avoid illegal
+ *   transitions, especially during destruction.
+ *
+ * **Usage Notes:**
+ *
+ * - Should be used in conjunction with `remember` and provided the `parentLifecycle` as a key to
+ *   ensure it updates correctly when the parent lifecycle changes.
+ * - By integrating with Compose's lifecycle, it ensures that resources are properly released when
+ *   the composable leaves the composition.
+ *
+ * @param parentLifecycle The parent [Lifecycle] with which this lifecycle owner should synchronize.
+ */
+internal class ComposeLifecycleOwner(
+  private val parentLifecycle: Lifecycle
+) : LifecycleOwner, RememberObserver {
+
+  private val registry = LifecycleRegistry(this)
+  override val lifecycle: Lifecycle
+    get() = registry
+
+  private val parentObserver = LifecycleEventObserver { _, event ->
+    // Any time the parent lifecycle changes state, perform the same change on our lifecycle.
+    registry.handleLifecycleEvent(event)
+  }
+
+  init {
+    // We do this right away to ensure that the parent lifecycle is observed as soon as possible
+    // to prevent any inconsistencies.
```

Comment:
We can be more specific: "We need to synchronize the lifecycles before the child ever even sees the lifecycle because compose's contract tries to guarantee that the lifecycle is in at least the `CREATED` state by the time composition is actually running. If we don't synchronize the lifecycles right away, then we break that invariant. One concrete case of this is that `SavedStateRegistry` requires its lifecycle to be `CREATED` before reading values from it, and consuming values from an SSR is a valid thing to do from composition directly, and in fact `AndroidComposeView` itself does this."